### PR TITLE
more robust way of loading resources for AppView sections

### PR DESCRIPTION
**Pull request by prydonius**
- stores data retrieved from WebSocket connections in Redux store
- wraps ServiceItems in a container which gets the relevant Service from the store
- ServiceItems executes an HTTP fetch for the Service to retrieve data
  quicker, as well as discover errors such as missing resources

refs #882

screen shot 2019-01-09 at 20 13 45

kubeapps-serviceitemloading

### CODE REVIEW

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx [hunk 1]
```diff
@@ -1,32 +1,73 @@
...
+  IKubeItem,
+  IResource,
+  IResourceRef,
+  IServiceSpec,
+  IServiceStatus,
+} from "../../../shared/types";
 
 interface IServiceItemProps {
-  service: IResource;
+  serviceRef: IResourceRef;
```
Comment (migmartri) [inline dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: Maybe serviceRef is redundant? Couldn't you get the name from `service: IKubeItem<IResource>;`?

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx [hunk 2]
```diff
@@ -1,32 +1,73 @@
...
-        </td>
+        <td>{serviceRef.name}</td>
+        {this.renderServiceInfo(service)}
       </tr>
     );
   }
 
-  private getExternalIP(): string {
-    const { service } = this.props;
+  private renderServiceInfo(service?: IKubeItem<IResource>) {
```
Comment (migmartri) [inline dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: I like this code, it's flow is quite easy to understand.

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.tsx [hunk 3]
```diff
@@ -38,10 +33,9 @@ class ServiceTable extends React.Component<IServiceTableProps> {
             </tr>
           </thead>
           <tbody>
-            {services.map(
-              s =>
-                s.item && <ServiceItem key={`services/${s.item.metadata.name}`} service={s.item} />,
-            )}
+            {serviceRefs.map(s => (
+              <ServiceItem key={`services/${s.namespace}/${s.name}`} serviceRef={s} />
```
Comment (migmartri) [inline dashboard/src/components/AppView/ServicesTable/ServicesTable.tsx]: `key={`services/${s.namespace}/${s.name}`}` related to my comment above. Wouldn't be great if serviceRef (s) could have a method to get this?

File: dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+function mapStateToProps({ kube }: IStoreState, props: IServiceItemContainerProps) {
+  const { serviceRef } = props;
+  const serviceKey = Kube.getResourceURL(
+    serviceRef.apiVersion,
+    "services",
+    serviceRef.namespace,
+    serviceRef.name,
+  );
+  return {
+    ...props,
```
Comment (migmartri) [inline dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: I'd be more specific on what we pass here. Just for clarity. 

I wrote it in another comment but I am not sure why you need to pass `serviceRef` to the component.

File: dashboard/src/containers/AppViewContainer/AppViewContainer.tsx [hunk 5]
```diff
@@ -31,6 +31,8 @@ function mapDispatchToProps(dispatch: ThunkDispatch<IStoreState, null, Action>)
     deleteApp: (releaseName: string, ns: string, purge: boolean) =>
       dispatch(actions.apps.deleteApp(releaseName, ns, purge)),
     getApp: (releaseName: string, ns: string) => dispatch(actions.apps.getApp(releaseName, ns)),
+    receiveResource: (payload: { key: string; resource: IResource }) =>
```
Comment (migmartri) [inline dashboard/src/containers/AppViewContainer/AppViewContainer.tsx]: This is temporary while we move the websockets handling outside of this component/container right? If that's the case I'd add a `TODO` note in this code and also inside the component.

File: dashboard/src/components/AppView/AppView.tsx [hunk 6]
```diff
@@ -129,23 +134,42 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
...
       return _.dropWhile(array, r => r.item && r.item.metadata.name === resource.metadata.name);
     };
+    let apiVersion = "v1";
+    let apiResource: string;
     switch (resource.kind) {
       case "Deployment":
         const newDeps = dropByName(this.state.deployments);
         newDeps.push(newItem);
         this.setState({ deployments: newDeps });
+        // We won't use resource.apiVersion as it may have a different version
```
Comment (andresmgot) [inline dashboard/src/components/AppView/AppView.tsx]: I am not sure I understand this comment, why don't we want to use the `apiVersion` in the cache?

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx (hunk 1 above)
Comment (andresmgot) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: `service` may be still empty, that's why serviceRef may be needed

File: dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import { IResourceRef, IStoreState } from "../../shared/types";
+
+interface IServiceItemContainerProps {
+  serviceRef: IResourceRef;
+}
+
+function mapStateToProps({ kube }: IStoreState, props: IServiceItemContainerProps) {
+  const { serviceRef } = props;
+  const serviceKey = Kube.getResourceURL(
+    serviceRef.apiVersion,
```
[the remark this answers is not in the document]
Comment (prydonius) [reply dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: Interesting idea, I think that would be quite nice actually

File: dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts (hunk 4 above)
Comment (prydonius) [reply dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: I use `serviceRef` in `ServiceItem` to display the name of the Service whilst fetching (the screenshot doesn't show it, but the name is shown right away, with a loading icon for the rest of the rows. It's not very important, but I like being able to see what resources are loading.

Comment (prydonius) [reply dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: screen shot 2019-01-10 at 13 09 13

Comment (prydonius) [reply dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: Now that I think about it, I can change this to just send a `name` prop to the ServiceItem component, since it doesn't need the full `serviceRef`!

File: dashboard/src/containers/AppViewContainer/AppViewContainer.tsx (hunk 5 above)
Comment (prydonius) [reply dashboard/src/containers/AppViewContainer/AppViewContainer.tsx]: Yep, good shout - I'll add a TODO!

File: dashboard/src/components/AppView/AppView.tsx (hunk 6 above)
Comment (prydonius) [reply dashboard/src/components/AppView/AppView.tsx]: So I need to do some testing here, my concern is that the API server will convert, for example an `extensions/v1beta1` to an `apps/v1` version for Deployments if available. So a chart manifest with an older API version will lead to inconsistencies. I think we need to fix the API version for fetching Deployments to `apps/v1` regardless of what the server returns or is in the release manifest to avoid this inconsistency.

Comment (prydonius) [reply dashboard/src/components/AppView/AppView.tsx]: On further testing, the API server doesn't actually change the apiVersion field on the object, so I'm able to consistently get `extensions/v1beta` back from the API. Given this, I think we can change this to use the `apiVersion` from the returned resource, though we should keep this in mind in case it can cause issues in the future.

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.tsx (hunk 3 above)
Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServicesTable.tsx]: Opted to just use the full resource URL here rather than creating another specific key, it has all we need to be unique and it doesn't matter if this key is verbose.

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx [hunk 8]
```diff
@@ -34,51 +33,55 @@ it("renders a message if there are no services or ingresses", () => {
...
-      .props(),
-  ).toMatchObject({ service: service1 });
-  expect(
-    wrapper
-      .find(ServiceItem)
-      .at(1)
-      .props(),
-  ).toMatchObject({ service: service2 });
-});
+// it("renders a table with a service with a LoadBalancer", () => {
```
Comment (migmartri) [inline dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx]: An alternative to commenting out the code would be to skip it (`xit` instead of `it`). At least you will remember to implement it.

File: dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+interface IServiceItemContainerProps {
+  serviceRef: ResourceRef;
+}
+
+function mapStateToProps({ kube }: IStoreState, props: IServiceItemContainerProps) {
+  const { serviceRef } = props;
+  return {
+    name: serviceRef.name,
+    service: kube.items[serviceRef.getResourceURL()],
```
Comment (migmartri) [inline dashboard/src/containers/ServiceItemContainer/ServiceItemContainer.ts]: Much cleaner now.

File: dashboard/src/shared/ResourceRef.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import { Kube } from "./Kube";
+import { IResource } from "./types";
+
+class ResourceRef {
+  // Creates a new ResourceRef instance from an existing IResource. Provide
+  // defaultNamespace to set if the IResource doesn't specify a namespace.
+  // TODO: add support for cluster-scoped resources, or add a ClusterResourceRef
+  // class.
+  public static newFromResource(r: IResource, defaultNamespace: string = "default") {
```
Comment (migmartri) [inline dashboard/src/shared/ResourceRef.ts]: why isn't this a proper class constructor? This sounds GO idiomatic.

File: dashboard/src/shared/ResourceRef.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
+import { Kube } from "./Kube";
+import { IResource } from "./types";
+
+class ResourceRef {
```
Comment (migmartri) [inline dashboard/src/shared/ResourceRef.ts]: I like it. I'd maybe add a comment explaining what this class means.

File: dashboard/src/components/AppView/AppView.tsx (hunk 6 above)
Comment (andresmgot) [reply dashboard/src/components/AppView/AppView.tsx]: +1 thanks for checking (BTW you forgot to remove the comment?)

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx (hunk 8 above)
Comment (andresmgot) [reply dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx]: do you plan to remove the tests?

Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx]: good point

File: dashboard/src/shared/ResourceRef.ts (hunk 10 above)
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: What if we want to create one without an IResource, in tests for example?

Comment (andresmgot) [reply dashboard/src/shared/ResourceRef.ts]: the pattern we have been following is to create a dummy object with the required elements that we reuse ass "default" props

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx (hunk 8 above)
Comment (andresmgot) [reply dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx]: but if the tests doesn't apply anymore, shoudln't we delete them? (as far as they are applied where they belong). I am not sure if I am missing something

File: dashboard/src/shared/ResourceRef.ts (hunk 10 above)
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: I think the issue is that that approach doesn't work for classes because it requires the `getResourceURL` to be present in that dummy object, but let me try again.

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx [hunk 12]
```diff
@@ -43,6 +53,9 @@ it("renders a view with IP", () => {
     spec: { ports: [{ port: 80 }], type: "LoadBalancer" },
     status: { loadBalancer: { ingress: [{ ip: "1.2.3.4" }] } },
   } as IResource;
-  const wrapper = shallow(<ServiceItem service={service} />);
+  kubeItem.item = service;
+  const wrapper = shallow(
+    <ServiceItem service={kubeItem} name={service.metadata.name} getService={jest.fn()} />,
+  );
   expect(wrapper).toMatchSnapshot();
 });
```
Comment (andresmgot) [inline dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx]: We would need to move here the tests to ensure that `"ItBehavesLike"` a loading component and that it renders an error if present.

Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx]: yep, although ItBehavesLike can't be used because we don't use `LoadingWrapper`, but rather use `LoadingSpinner` directly.

File: dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx (hunk 8 above)
Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServicesTable.test.tsx]: Yes, I just commented them out temporarily because it wasn't compiling in the dev environment

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx (hunk 12 above)
Comment (andresmgot) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx]: ah, I didn't notice that. Could you add a comment there explaining why do you use that instead of the LoadingWrapper? (Since we usually use that by default).

Also then here we would need to create more verbose tests just checking if the LoadingSpinner exists,

Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.test.tsx]: Good shout, and after discussing with @migmartri I've changed this to use LoadingWrapper. Also created https://github.com/kubeapps/kubeapps/issues/925 for ConfirmDialog which is using LoadingSpinner directly. Thanks!

File: dashboard/src/components/AppView/AppView.tsx (hunk 6 above)
Comment (prydonius) [reply dashboard/src/components/AppView/AppView.tsx]: done, thanks!

File: dashboard/src/shared/ResourceRef.ts (hunk 10 above)
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: I switched this to a proper constructor - thanks!

File: dashboard/src/shared/ResourceRef.ts (hunk 11 above)
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: done

**Comment by prydonius** [comment]
@andresmgot @migmartri I believe I've addressed all the review feedback now and added some more test coverage - PTAL!

File: dashboard/src/shared/ResourceRef.test.ts [hunk 13]
```diff
@@ -0,0 +1,78 @@
+import { Kube } from "./Kube";
+import ResourceRef from "./ResourceRef";
+import { IResource } from "./types";
+
+describe("ResourceRef", () => {
+  describe("constructor", () => {
+    it("it returns a ResourceRef with the correct details", () => {
+      const r = {
```
Comment (migmartri) [inline dashboard/src/shared/ResourceRef.test.ts]: I was not that difficult in the end wasn't it?

File: dashboard/src/components/AppView/AppView.tsx [hunk 14]
```diff
@@ -231,6 +262,7 @@ class AppView extends React.Component<IAppViewProps, IAppViewState> {
           break;
         case "Service":
           result.services.push(resource);
+          result.serviceRefs.push(new ResourceRef(resource.item, namespace));
```
Comment (migmartri) [inline dashboard/src/components/AppView/AppView.tsx]: So the namespace in this case might not be inside the resourceItem? What's `namespace` in this case?

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx [hunk 15]
```diff
@@ -1,32 +1,67 @@
...
+    if (service === undefined || service.isFetching) {
+      return (
+        <td colSpan={4}>
+          <LoadingWrapper size="inline-small" />
```
Comment (migmartri) [inline dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: So this is going to be showing a loading message per element being loaded? It might be too much if we extrapolate this to let's say deployments no?

Do we need to show loading at all here? I thought that we already had some kind of struct that indicated that at least one resource was being loaded and we used that in the whole table?

File: dashboard/src/components/AppView/AppView.tsx (hunk 14 above)
Comment (prydonius) [reply dashboard/src/components/AppView/AppView.tsx]: The release namespace is used when the resource in the manifest doesn't have a namespace (likely since charts leave the namespace undefined so it deploys in whatever namespace the user chooses). I'll rename this to `releaseNamespace`, didn't want to before since it was already that.

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx (hunk 15 above)
Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: Correct, this was intended, we now show the name of the resource and a loading icon next to it. Each resource is loaded independently in parallel so it's possible that some are still loading whilst others have data to show.

> It might be too much if we extrapolate this to let's say deployments no?

What will be too much? In most cases it loads pretty fast and I don't think the spinners are too annoying (that said I haven't seen what this looks like on a chart with a lot of Services).

> Do we need to show loading at all here? I thought that we already had some kind of struct that indicated that at least one resource was being loaded and we used that in the whole table?

The table used to ensure all Services were loaded before displaying anything, but now it just renders `ServiceItem` containers. We could wrap the ServiceTable into a container and have a single loader, but I actually _like_ that we show each Service loading independently, it feels more responsive to me.

Comment (andresmgot) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: For me it's okay to have a loading message per element. If having so many spinners is annoying we can remove it and have just a "Loading..." message.

File: dashboard/src/shared/ResourceRef.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  public apiVersion: string;
+  public kind: string;
+  public name: string;
+  public namespace: string;
+
+  // Creates a new ResourceRef instance from an existing IResource. Provide
+  // defaultNamespace to set if the IResource doesn't specify a namespace.
+  // TODO: add support for cluster-scoped resources, or add a ClusterResourceRef
+  // class.
+  constructor(r: IResource, defaultNamespace: string = "default") {
```
[the remark this answers is not in the document]
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: Hmm good question, the issue here is that if the IResource defines a namespace in `metadata.namespace`, that will always be used. The `defaultNamespace` is only used if `metadata.namespace` is undefined. If you know that the IResource you're passing to `ResourceRef` contains the namespace, you don't need to provide the `defaultNamespace`, if you did, you'd have to set it to `resource.metadata.namespace` which would be a bit annoying.

I do get that this is a bit confusing though, as if `metadata.namespace` is undefined, you probably will need to provide `defaultNamespace`, it's unlikely you'd want that to be "default".

Another option here is to have the signature be `defaultNamespace?: string`, and then in the code we would need to throw an error if both `metadata.namespace` and `defaultNamespace` are undefined. However, I'm not sure it's worth doing that.

File: dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx (hunk 15 above)
Comment (prydonius) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: This is what it looks like, I quite like it. IMO it's subtle and not that annoying.

kubeapps-serviceitemloading

Comment (migmartri) [reply dashboard/src/components/AppView/ServicesTable/ServiceItem.tsx]: It does not look bad actually. I still have a preference for a global (per table) loading spinner but I understand that it complicates and couples the code, so I'd not push for it.

Just being picky and since asking for things is free. Something that annoys me is the alignment of the spinners. Something very slick would be the use of blurry lines a.l.a Facebook or Slack https://kyusuf.com/post/fake-it-til-you-make-it-css/

File: dashboard/src/shared/ResourceRef.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (andresmgot) [reply dashboard/src/shared/ResourceRef.ts]: I guess it's better then to throw the error if `metadata.namespace` and `defaultNamespace` are unset. It could be quite confusing and difficult to debug if suddenly there are unexpected requests to the `default` namespace.

[the remark this answers is not in the document]
Comment (prydonius) [reply dashboard/src/shared/ResourceRef.ts]: done, and added a test - thanks!
